core/src/test/java/org/apache/calcite/util/SealableMapTest.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for the seal/unseal functionalities of {@link SealableMap}
+ */
+public class SealableMapTest {
+  //~ Constructors -----------------------------------------------------------
+
+  public SealableMapTest() {
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Test public void testSealUnseal() {
+    Map<String, String> backingMap = new HashMap<>();
+    SealableMap<String, String> sealableMap = new SealableMap<>(backingMap);
+    sealableMap.put("foo1", "bar1");
+    sealableMap.put("foo2", "bar2");
+
+    // start off unsealed
+    ensureUnsealed(sealableMap, backingMap);
+
+    // seal and verify
+    sealableMap.seal();
+    ensureSealed(sealableMap, backingMap);
+
+    // unseal and verify
+    sealableMap.unseal();
+    ensureUnsealed(sealableMap, backingMap);
+  }
+
+  @Test public void testEqualsHashCode() {
+    Map<String, String> backingMap = new HashMap<>();
+    backingMap.put("foo", "bar");
+
+    // will be equal
+    SealableMap<String, String> map1 = new SealableMap<>(backingMap);
+    SealableMap<String, String> map2 = new SealableMap<>(backingMap);
+
+    // will be unequal because this map will be sealed
+    SealableMap<String, String> map3 = new SealableMap<>(backingMap);
+    map3.seal();
+
+    // will be unequal because map entries are different
+    SealableMap<String, String> map4 = new SealableMap<>(new HashMap<>());
+    map4.put("foo", "baz");
+
+    assertEquals(map1, map2);
+    assertEquals(map1.hashCode(), map2.hashCode());
+
+    assertNotEquals(map1, map3);
+    assertNotEquals(map1, map4);
```

Comment:
please also assert map3 and map4 are different

Reply:
good point, done !